fspiop-api/documents/Tracing v1.0.md

```diff
@@ -0,0 +1,221 @@
+
+> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`00f067aa0ba902b7`-01<br/>
+> tracestate: `moja`=`00f067aa0ba902b7`,fsp1=t61rcWkgMzE<br/>
+
+<br/>
+4.1.3. `FSP2` receives the request  and responds with a `HTTP 202` accepted. The funds are committed and `FSP2` sends a **PUT /transfers** callback including the following mutated trace information to the `Mojaloop Switch`:<br/><br/>
+
+> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`b9c7c989f97918e1`-01<br/>
+> tracestate: `fsp2`=`ucfJifl5GOE`,moja=00f067aa0ba902b7,fsp1=t61rcWkgMzE<br/>
+
+<br/>
+4.1.4. The `Mojaloop Switch` receives the **PUT /transfers** and responds with a `HTTP 200`. The `Mojaloop Switch` will reconstuct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the the **PUT /transfers** back to `FSP1`:<br/><br/>
+
+> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`53ce929d0e0e4736`-01<br/>
+> tracestate: `moja`=`53ce929d0e0e4736`,fsp2=ucfJifl5GOE,fsp1=t61rcWkgMzE<br/>
+
+<br/>
+4.1.5. `FSP1` receives the fulfil response responding with an `HTTP 200`, and completes the transfer.
+
+#### 4.2. POST /transfers request being sent by non-trace-participating FSP1 to non-trace-participating FSP2 through a Mojaloop Switch
+
+This will example will result in two disparate end-to-end trace graphs being created which are NOT linked as follows:
+> 4.2.a. `FSP1` to `FSP2` being routed through a `Mojaloop Switch` for a  **POST /transfers**<br/>
+> 4.2.b. `FSP2` to `FSP1` being routed through a `Mojaloop Switch` for a  **PUT /transfers** callback<br/>
+
+<br/>
+4.2.1. `FSP1` sends a POST /transfers to a `Mojaloop Switch` which includes no trace information.
+
+4.2.2. The `Mojaloop Switch` receives the POST /transfers and responds with a HTTP 202 accepted. The `Mojaloop Switch` generates a new trace as none was found in the request headers before forwarding the the POST /transfers to `FSP2`:<br/><br/>
+
+> traceparent: `00`-`0af7651916cd43dd8448eb211c80319c`-`00f067aa0ba902b7`-`01`<br/>
+> tracestate: `moja`=`00f067aa0ba902b7`<br/>
+
+<br/>
+4.2.3. `FSP2` receives the request, commits the funds and responds with a PUT /transfers callback forwarding the unmutated trace information:<br/><br/>
+
+> traceparent: 00-0af7651916cd43dd8448eb211c80319c-b9c7c989f97918e1-01 <br/>
+> tracestate: moja=00f067aa0ba902b7
+
+<br/>
+4.2.4. The `Mojaloop Switch` receives the PUT /transfers and responds with a HTTP 200. The `Mojaloop Switch` will reconstruct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the the PUT /transfers back to `FSP1`:<br/><br/>
```

Comment:
Can you please explain why the Mojaloop Switch should reconstruct the trace here? Seems a bit unnecessary as it is the only participant in the trace, but I will assume that there is a good reason :)

```suggestion
4.2.4. The `Mojaloop Switch` receives the **PUT /transfers** and responds with a HTTP 200. The `Mojaloop Switch` will reconstruct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the **PUT /transfers** back to `FSP1`:<br/><br/>
```

Reply:
By this, I mean "reconstruct" the trace context within the switch. I have rewarded it as such.

It is required because each time the trace is processed by a microservices's (at a minimum - as we could even go down into internal function calls, etc), a "span" must be created to represent the work of that microservice.

Does that make it clear?